x/ccv/provider/keeper/key_assignment.go

```diff
@@ -45,58 +45,43 @@ func (k Keeper) SetValidatorConsumerPubKey(
 	store.Set(types.ConsumerValidatorsKey(chainID, providerAddr), bz)
 }
 
-// IterateValidatorConsumerPubKeys iterates over the validators public keys assigned for a consumer chain
-func (k Keeper) IterateValidatorConsumerPubKeys(
-	ctx sdk.Context,
-	chainID string,
-	cb func(providerAddr sdk.ConsAddress, consumerKey tmprotocrypto.PublicKey) (stop bool),
-) {
+// GetAllValidatorConsumerPubKeys gets all the validators public keys assigned for a consumer chain
+// If chainID is nil, it returns all the validators public keys assigned for all consumer chains
+//
+// Note that the validators public keys assigned for a consumer chain are stored under keys
+// with the following format: UnbondingOpIndexBytePrefix | len(chainID) | chainID | providerAddress
+// Thus, the returned array is
+//   - in ascending order of providerAddresses, if chainID is not nil;
+//   - in undetermined order, if chainID is nil.
+func (k Keeper) GetAllValidatorConsumerPubKeys(ctx sdk.Context, chainID *string) (validatorConsumerPubKeys []types.ValidatorConsumerPubKey) {
 	store := ctx.KVStore(k.storeKey)
-	prefix := types.ChainIdWithLenKey(types.ConsumerValidatorsBytePrefix, chainID)
-	iter := sdk.KVStorePrefixIterator(store, prefix)
-	defer iter.Close()
-	for ; iter.Valid(); iter.Next() {
-		_, providerAddr, err := types.ParseChainIdAndConsAddrKey(types.ConsumerValidatorsBytePrefix, iter.Key())
-		if err != nil {
-			panic(err)
-		}
-		var consumerKey tmprotocrypto.PublicKey
-		err = consumerKey.Unmarshal(iter.Value())
-		if err != nil {
-			panic(err)
-		}
-		stop := cb(providerAddr, consumerKey)
-		if stop {
-			break
-		}
+	var prefix []byte
+	if chainID == nil {
```

Comment:
Could we add a comment on what's going on with prefix here. Why would we allow chainID to be nil?